File: src/converter/plugins/InheritedTypeLiteralPlugin.ts
import ts, {IntersectionTypeNode, MappedTypeNode, TypeLiteralNode, TypeReferenceNode} from "typescript";
import {ifPresent, Render} from "../render.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";
import {InheritanceModifierService, inheritanceModifierServiceKey} from "./InheritanceModifierPlugin.js";
import {createAnonymousDeclarationPlugin} from "./AnonymousDeclarationPlugin.js";
import {extractTypeParameters, renderDeclaration, renderReference} from "../extractTypeParameters.js";
import {ConverterContext} from "../context.js";
import {InjectionService, injectionServiceKey} from "./InjectionPlugin.js";
import {convertMappedTypeBody} from "./MappedTypePlugin.js";
import {convertTypeLiteralBody} from "./TypeLiteralPlugin.js";
import {InjectionType} from "../injection.js";

export function isInheritedTypeLiteral(node: ts.Node): node is IntersectionTypeNode {
    return (
        ts.isIntersectionTypeNode(node)
        && node.types.every(it => (
            ts.isTypeReferenceNode(it)
            || ts.isTypeLiteralNode(it)
            || ts.isMappedTypeNode(it)
        ))
    )
}

export function convertInheritedTypeLiteral(
    node: IntersectionTypeNode,
    name: string,
    typeParameters: string | undefined,
    context: ConverterContext,
    render: Render,
) {
    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    const inheritanceModifierService = context.lookupService<InheritanceModifierService>(inheritanceModifierServiceKey)
    const injectionService = context.lookupService<InjectionService>(injectionServiceKey)

    const inheritanceModifier = inheritanceModifierService?.resolveInheritanceModifier(node, context)
    const injections = injectionService?.resolveInjections(node, InjectionType.MEMBER, context, render)
    const heritageInjections = injectionService?.resolveInjections(node, InjectionType.HERITAGE_CLAUSE, context, render)

    const typeReferences = node.types.filter((it): it is TypeReferenceNode => ts.isTypeReferenceNode(it))
    const typeLiterals = node.types.filter((it): it is TypeLiteralNode => ts.isTypeLiteralNode(it))
    const mappedType = node.types.find((it): it is MappedTypeNode => ts.isMappedTypeNode(it))

    const heritageTypes = typeReferences
        .map(type => render(type))
        .filter(Boolean)
        .join(", ")

    const injectedHeritageClauses = heritageInjections
        ?.filter(Boolean)
        ?.join(", ")

    const fullHeritageClauses = [heritageTypes, injectedHeritageClauses]
        .filter(Boolean)
        .join(", ")

    const members = typeLiterals
        .map(it => convertTypeLiteralBody(it, context, render))
        .join("\n")

    let accessors = ""

    if (mappedType) {
        accessors = convertMappedTypeBody(mappedType, context, render)
    }

    const injectedMembers = (injections ?? [])
        .join("\n")

    return `
${ifPresent(inheritanceModifier, it => `${it} `)}external interface ${name}${ifPresent(typeParameters, it => `<${it}> `)}${(ifPresent(fullHeritageClauses, it => ` : ${it}`))} {
${ifPresent(accessors, it => `${it}\n`)}${members}${ifPresent(injectedMembers, it => `\n${it}`)}
}
    `.trim()
}

export const inheritedTypeLiteralPlugin = createAnonymousDeclarationPlugin(
    (node, context, render) => {
        if (!isInheritedTypeLiteral(node)) return null

        const name = context.resolveName(node)

        const typeParameters = extractTypeParameters(node, context)

        const declaration = convertInheritedTypeLiteral(node, name, renderDeclaration(typeParameters, render), context, render)

        const reference = `${name}${ifPresent(renderReference(typeParameters, render), it => `<${it}>`)}`

        return {name, declaration, reference};
    }
)

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/InheritedTypeLiteralPlugin.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.configuration.NamespaceStrategy
import io.github.sgrishchenko.karakum.extension.*
import typescript.*
import kotlin.contracts.contract

@Suppress("CANNOT_CHECK_FOR_EXTERNAL_INTERFACE")
fun isInheritedTypeLiteral(node: Node): Boolean {
    contract {
        returns(true) implies (node is IntersectionTypeNode)
    }

    return isIntersectionTypeNode(node) && node.types.asArray().all {
        isTypeReferenceNode(it)
                || isTypeLiteralNode(it)
                || isMappedTypeNode(it)
    }
}

fun convertInheritedTypeLiteral(
    node: IntersectionTypeNode,
    name: String,
    typeParameters: String?,
    isInlined: Boolean,
    context: Context,
    render: Render<Node>,
): String {
    val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
    val namespaceInfoService = context.lookupService<NamespaceInfoService>(namespaceInfoServiceKey)
    val inheritanceModifierService = context.lookupService<InheritanceModifierService>(inheritanceModifierServiceKey)
    val injectionService = context.lookupService<InjectionService>(injectionServiceKey)

    val inheritanceModifier = inheritanceModifierService?.resolveInheritanceModifier(node, context)
    val injections = injectionService?.resolveInjections(node, InjectionType.MEMBER, context, render)
    val heritageInjections = injectionService?.resolveInjections(node, InjectionType.HERITAGE_CLAUSE, context, render)

    val namespace = typeScriptService?.findClosestNamespace(node)

    var externalModifier = "external "

    if (isInlined && namespace != null && namespaceInfoService?.resolveNamespaceStrategy(namespace) == NamespaceStrategy.`object`) {
        externalModifier = ""
    }

    val typeReferences = node.types.asArray().filter { isTypeReferenceNode(it) }
    val typeLiterals = node.types.asArray().filter { isTypeLiteralNode(it) }
    val mappedType = node.types.asArray().find { isMappedTypeNode(it) }

    val heritageTypes = typeReferences
        .map { render(it) }
        .filter { it.isNotEmpty() }
        .joinToString(separator = ", ")

    val injectedHeritageClauses = heritageInjections
        ?.filter { it.isNotEmpty() }
        ?.joinToString(separator = ", ")

    val fullHeritageClauses = arrayOf(heritageTypes, injectedHeritageClauses ?: "")
        .filter { it.isNotEmpty() }
        .joinToString(separator = ", ")

    val members = typeLiterals
        .mapNotNull {
            if (!isTypeLiteralNode(it)) return@mapNotNull null
            convertTypeLiteralBody(it, context, render)
        }
        .joinToString(separator = "\n")

    var accessors = ""

    if (mappedType != null && isMappedTypeNode(mappedType)) {
        accessors = convertMappedTypeBody(mappedType, context, render)
    }

    val injectedMembers = (injections ?: emptyArray()).joinToString(separator = "\n")

    return """
${ifPresent(inheritanceModifier) { "$it "}}${externalModifier}interface ${name}${ifPresent(typeParameters) { "<${it}> "}}${(ifPresent(fullHeritageClauses) { " : $it"})} {
${ifPresent(accessors) { "${it}\n" }}${members}${ifPresent(injectedMembers) { "\n${it}"}}
}
    """.trim()
}

fun createInheritedTypeLiteralPlugin() = createAnonymousDeclarationPlugin plugin@{ node, context, render ->
    if (!isInheritedTypeLiteral(node)) return@plugin null

    val name = context.resolveName(node)

    val typeParameters = extractTypeParameters(node, context)

    val declaration =
        convertInheritedTypeLiteral(node, name, renderDeclaration(typeParameters, render), false, context, render)

    val reference = "${name}${ifPresent(renderReference(typeParameters, render)) { "<${it}>" }}"

    AnonymousDeclaration(
        name = name,
        declaration = declaration,
        reference = reference
    )
}
